core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java

```diff
@@ -135,22 +137,21 @@ public String executor() {
 
     /**
      * Executed on the node that should be running the task to find and return the running task. Falls back to
-     * {@link #getFinishedTaskFromIndex(Task, GetTaskRequest, ActionListener)} if the task isn't still running.
+     * {@link #getFinishedTaskFromIndex(Task, GetTaskRequest, TaskInfo, ActionListener)} if the task isn't still running.
      */
     void getRunningTaskFromNode(Task thisTask, GetTaskRequest request, ActionListener<GetTaskResponse> listener) {
         Task runningTask = taskManager.getTask(request.getTaskId().getId());
         if (runningTask == null) {
-            getFinishedTaskFromIndex(thisTask, request, listener);
+            getFinishedTaskFromIndex(thisTask, request, null, listener);
         } else {
             if (request.getWaitForCompletion()) {
                 // Shift to the generic thread pool and let it wait for the task to complete so we don't block any important threads.
                 threadPool.generic().execute(new AbstractRunnable() {
                     @Override
                     protected void doRun() throws Exception {
                         taskManager.waitForTaskCompletion(runningTask, waitForCompletionTimeout(request.getTimeout()));
-                        // TODO look up the task's result from the .tasks index now that it is done
-                        listener.onResponse(
-                                new GetTaskResponse(new PersistedTaskInfo(runningTask.taskInfo(clusterService.localNode(), true))));
+                        TaskInfo snapshotOfRunningTask = runningTask.taskInfo(clusterService.localNode(), true);
+                        getFinishedTaskFromIndex(thisTask, request, snapshotOfRunningTask, listener);
```

Comment:
It feels like `getFinishedTaskFromIndex` is doing too much right now. I think it might be cleaner to leave it as it was and instead handle it right here by wrapping the listener and capturing the `ResourceNotFoundException` exception in `onFailure`. You will be able to run taskInfo in a lazy fashion as well in this case. What do you think?